File: Assignment1/sfs/heap.py
class Heapobject(object):
	def __init__(self, key):
		super(Heapobject, self).__init__()
		self.key = key

def parent(index):
	return (index-1)/2

def left(index):
	return 2*index + 1

def right(index):
	return 2*index + 2

def insertKey(heap, key):
	heap.append(Heapobject(key))
	i = len(heap) - 1
	while i > 0 and heap[parent(i)].key > heap[i].key:
		heap[parent(i)] , heap[i] = heap[i] , heap[parent(i)]
		i = parent(i)
	return heap

def minHeapify(heap, i):
	l = left(i)
	r = right(i)
	heapsize = len(heap)
	if l < heapsize and heap[l].key < heap[i].key :
		largest = l
	else:
		largest = i
	if r < heapsize and heap[r].key < heap[largest].key :
		largest = r

	if largest != i :
		heap[i], heap[largest] = heap[largest], heap[i]
		minHeapify(heap, largest)

def extractMin(heap):
	heapsize = len(heap)
	if heapsize < 1:
		return -1
	minimum = heap[0]
	heap[0] = heap[heapsize - 1 ] 
	del heap[-1]

	minHeapify(heap, 0)
	return minimum

if __name__ == '__main__':
	heap = []
	insertKey(heap,16)
	insertKey(heap,14)
	insertKey(heap,10)
	insertKey(heap,9)
	insertKey(heap,1)
	insertKey(heap,2)
	insertKey(heap,3)
	print heap
	print extractMin(heap).key
	print extractMin(heap).key
	print extractMin(heap)
	print extractMin(heap)
	print extractMin(heap)
	print extractMin(heap)
	print extractMin(heap)
	print heap

File: Assignment1/sfs/heap.cpp
#include <bits/stdc++.h>
using namespace std;

int heap[1001];
int len = 0 ;
int man = 0 ; 

int parent(int idx){
	return (idx-1)/2;
}

int left(int idx){
	return 2*idx + 1;
}

int right(int idx){
	return 2*idx + 2;
}

void swap(int a, int b){
	int temp = heap[a];
	heap[a] = heap[b];
	heap[b] = temp;

	if(man == a){
		man = b;
	}
	else if(man == b){
		man = a;
	}

}

void insertKey(int key){
	heap[len] = key;
	int i = len;
	len++;
	while( i > 0 && heap[parent(i)] > heap[i]){
		swap(parent(i), i );
		i = parent(i);
	}
}

int minHeapify(int i){
	int l = left(i);
	int r = right(i);
	int largest;
	if (l < len && heap[l] < heap[i]){
		largest = l;
	}else{
		largest = i;
	}
	if(r < len && heap[r] < heap[largest]){
		largest = r;
	}

	if(largest != i){
		swap(i, largest);
		minHeapify(largest);
	}
}

int decreaseKey(){
	heap[0]--;
	minHeapify(0);
}

int increaseKey(){
	heap[man]++;
	int i = man;
	while(i > 0 && heap[parent(i)] < heap[i]){
		swap(parent(i), i);
		i = parent(i);
	}
}

int main(){
	int n, temp, l;
	scanf("%d",&n);
	scanf("%d",&temp);
	l = temp;
	insertKey(temp);
	for (int i = 1; i < n; ++i)
	{
		scanf("%d",&temp);
		if(temp >= l) insertKey(temp);
	}
	int count = 0;
	while(man != 0){
		decreaseKey();
		increaseKey();
		count++;
	}
	if(len > 2 && (heap[0] == heap[1] || heap[0] == heap[2]))
	{
		count++;
	}
	else if(len > 1 && heap[0] == heap[1]  ){
		count++;
	}
	cout << len << endl;
}